driver/nllloss_driver.hpp

```diff
@@ -0,0 +1,340 @@
+
+    int status;
+
+    for (int i = 0; i < in_sz; i++)
+    {
+        in[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(-10.0), static_cast<Tgpu>(-(1e-2)));
+    }
+    status = in_dev->ToGPU(q, in.data());
+
+    for (int i = 0; i < target_sz; i++)
+    {
+        target[i] = prng::gen_A_to_B<int>(static_cast<int>(0), static_cast<int>(C-1));
+    }
+    status |= target_dev->ToGPU(q, target.data());
+
+    for (int i = 0; i < weight_sz; i++)
+    {
+        weight[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(-10.0), static_cast<Tgpu>(10.0));
+    }
+    status |= weight_dev->ToGPU(q, weight.data());
+
+    status |= out_dev->ToGPU(q, out.data());
+
+    if(status != 0)
+        std::cout << "Error copying data to GPU\n" << std::endl;
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int NLLLossDriver<Tgpu, Tref>::RunForwardGPU()
+{
+    float kernel_total_time = 0.0;
+    float kernel_first_time = 0.0;
+
+    Timer t;
+    START_TIME
+
+    for(int i = 0; i < inflags.GetValueInt("iter"); i++)
+    {
+        miopenNLLLossForward(GetHandle(),
+                             inputDesc,
+                             in_dev->GetMem(),
+                             targetDesc,
+                             target_dev->GetMem(),
+                             weightDesc,
+                             weight_dev->GetMem(),
+                             outputDesc,
+                             out_dev->GetMem(),
+                             ignore_index);
+
+        float time = 0.0;
+        miopenGetKernelTime(GetHandle(), &time);
+        kernel_total_time += time;
+        if(i == 0)
+            kernel_first_time = time;
+    }
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        STOP_TIME
+        int iter = inflags.GetValueInt("iter");
+        if(WALL_CLOCK)
+            printf("Wall-clock Time Forward NLLLoss Elapsed: %f ms\n", t.gettime_ms() / iter);
+
+        float kernel_average_time =
+            iter > 1 ? (kernel_total_time - kernel_first_time) / (iter - 1) : kernel_first_time;
+        printf("GPU Kernel Time Forward NLLLoss Elapsed: %f ms\n", kernel_average_time);
+    }
+
+    out_dev->FromGPU(GetStream(), out.data());
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int NLLLossDriver<Tgpu, Tref>::RunForwardCPU()
+{
+    mloNLLLossForwardRunHost<Tgpu, Tref>(inputDesc,
+                                         in.data(),
+                                         target.data(),
+                                         weight.data(),
+                                         out_host.data(),
+                                         ignore_index);
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int NLLLossDriver<Tgpu, Tref>::RunBackwardGPU()
+{
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+Tref NLLLossDriver<Tgpu, Tref>::GetTolerance()
```

Comment:
Refer to https://github.com/ROCm/MIOpen/pull/2833 and modify the tolerance value.

Reply:
changed as per your comment
```c++
template <typename Tgpu, typename Tref>
Tref NLLLossDriver<Tgpu, Tref>::GetTolerance()
{
    // Computation error of fp16 is ~2^13 (=8192) bigger than
    // the one of fp32 because mantissa is shorter by 13 bits.
    auto tolerance = std::is_same<Tgpu, float>::value ? 1.5e-6 : 8.2e-3;

    // bf16 mantissa has 7 bits, by 3 bits shorter than fp16.
    if(std::is_same<Tgpu, bfloat16>::value)
        tolerance *= 8.0;
    return tolerance;
}
```